### PR TITLE
entropy adjustments: check return value of rdrand/rdseed (and retry up to 10 times), cope with AMD ryzen 3000 bug

### PR DESCRIPTION
**Pull request by hannesm**
check rdrand and rdseed return values (CR=1), as recommended by Intel (https://software.intel.com/en-us/articles/intel-digital-random-number-generator-drng-software-implementation-guide)

detect bad rdrand on ryzen 3000 (which returns all-ones) and disable there

### CODE REVIEW

File: src/native/entropy_cpu_stubs.c [hunk 1]
```diff
@@ -59,17 +59,27 @@ enum cpu_rng_t {
...
     __cpuid (1, eax, ebx, ecx, edx);
-    if (ecx & bit_RDRND) __cpu_rng = RNG_RDRAND;
+    if (ecx & bit_RDRND) {
+      /* AMD Ryzen 3000 bug where RDRAND always returns 0xFFFFFFFF */
+      for (int i = 0; i < RETRIES; i++)
+        if (_rdrand_step(&r) == 1 && r != (random_t) (-1)) {
+          __cpu_rng = RNG_RDRAND;
+          break;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (hannesm) [reply src/native/entropy_cpu_stubs.c]: thanks for the pointer, I pushed 548dbfb to address your concerns (rdseed is checked to not always return -1)

File: src/native/entropy_cpu_stubs.c [hunk 2]
```diff
@@ -91,12 +101,14 @@ CAMLprim value caml_cycle_counter (value __unused(unit)) {
...
+  for (int i = 0; i < RETRIES; i++) {
+    if (__cpu_rng == RNG_RDSEED) {
+      if (_rdseed_step (&r) == 1) return Val_long (r);
+    } else if (__cpu_rng == RNG_RDRAND) {
+      if (_rdrand_step (&r) == 1) return Val_long (r);
+    }
   }
-  return Val_long (r); /* Zeroed-out if carry == 0. */
+  return Val_long (0);
 #else
```
[the remark this answers is not in the document]
Comment (hannesm) [reply src/native/entropy_cpu_stubs.c]: TL;DR: not worth checking / retrying in this function (49a5a25 reverts that) -- but for initial bootstrap/entropy harvesting use checked rdrand/rdseed multiple times (same commit) -- hope this addresses your concerns (and thanks for the `do .. while` code and switch, I used it :)

File: src/native/entropy_cpu_stubs.c [hunk 3]
```diff
@@ -98,17 +105,50 @@ CAMLprim value caml_cycle_counter (value __unused(unit)) {
...
+    break;
+  case RNG_RDRAND:
+    do { ok = _rdrand_step (&r); } while ( !(ok | !--i) );
+    break;
+  case RNG_NONE:
+    break;
   }
+  return Val_long(r);
+#else
+  /* ARM: CPU-assisted randomness here. */
```
Comment (cfcs) [inline src/native/entropy_cpu_stubs.c]: Not for now, but in the future ARM is getting RDSEED-like instructions (don't believe there are currently any chips available with this 8.5-A ISA):
https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/arm-a-profile-architecture-2018-developments-armv85a
>  Armv8.5-A also contains other small features, including:
> Random Number instructions – providing Deterministic and True Random Numbers conforming to various National and International Standards.
Maybe worth leaving as a comment to future readers?

Comment (cfcs) [reply src/native/entropy_cpu_stubs.c]: Put this in https://github.com/mirage/mirage-crypto/issues/19

**Pull request merged**